### ISSUE TITLE
Can't iterate over iterators

### ISSUE THREAD

**Issue by zr40** [lfairy/maud#14]
The following iteration works:

```rust
let vec: Vec<()> = vec!();
html! {
    $for _ in &vec {
    }
};
```

However, iterating over `vec.iter()` fails to compile:

```rust
let vec: Vec<()> = vec!();
let iter = vec.iter();
html! {
    $for _ in &iter {
    }
};
```

This fails with:
```
error: the trait `core::iter::Iterator` is not implemented for the type `&core::slice::Iter<'_, ()>` [E0277]
```

Iterating over `vec.into_iter()` fails to compile with a similar error.

**Comment by lfairy** [comment]
Take out that extra `&`.

**Comment by zr40** [comment]
Without the `&`, it also fails to compile:

```
error: cannot move out of captured outer variable in an `Fn` closure
```

(might be related to issue #13)

**Comment by lfairy** [comment]
An `html!` block is a `Fn` closure, so everything it refers to must be borrowed not moved. This is why neither `.into_iter()` nor passing in an externally-constructed iterator work. Both of these require moving out of an outer variable, which `Fn` closures do not allow.

I'm open to changing this behavior, though. Do you have any use cases for consuming a structure in the template rather than borrowing it?

I might drop the closure machinery altogether, and expose only a `write!`-style macro instead. The current design is based on `format_args!` but I don't see anyone taking advantage of this flexibility.

**Comment by lfairy** [comment]
> Do you have any use cases for consuming a structure in the template rather than borrowing it?

It appears that there are many such use cases in the wild.

For example, if someone wants to stream the contents of a file line-by-line, they'd use [`BufRead::lines`](http://doc.rust-lang.org/nightly/std/io/trait.BufRead.html#method.lines). But this method consumes the original file.

I'll keep this issue closed as #18 sums up the issue more clearly. Either way I think the issues you have are valid.
